### PR TITLE
Use explicit crypto protos

### PR DESCRIPTION
**Pull request by ipetr0v**
This PR makes existing protos use explicit `EncryptedRequest` and `EncryptedResponse` protos and updates the corresponding logic.

Updated logic includes:
- `messages.proto` uses explicit protos
- `oak_functions.proto` uses explicit protos
- Updates Rust, C++ and Java client and transport logic
  - No more serialization/deserialization
- Updates Rust, C++ and Java encryptors
  - They now accept explicit proto messages
- Updates OakFunctions on Restricted Kernel and Oak Containers
  - Channel between the Launcher and the Application now uses explicit protos
- Updates all corresponding tests

This is a BREAKING CHANGE and will require updates in the internal code!

Fixes https://github.com/project-oak/oak/issues/4037
Fixes https://github.com/project-oak/oak/issues/4267

### CODE REVIEW

File: java/src/main/java/com/google/oak/crypto/DecryptionResult.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package com.google.oak.crypto;
+
+public final class DecryptionResult {
```
Comment (tiziano88) [inline java/src/main/java/com/google/oak/crypto/DecryptionResult.java:19]: Shouldn't this be used as input to the encrypt function too? (and renamed)

File: java/src/main/java/com/google/oak/crypto/Encryptor.java [hunk 2]
```diff
@@ -16,6 +16,8 @@
 
 package com.google.oak.crypto;
 
+import com.google.oak.crypto.v1.EncryptedRequest;
+import com.google.oak.crypto.v1.EncryptedResponse;
```
Comment (tiziano88) [inline java/src/main/java/com/google/oak/crypto/Encryptor.java:20]: Why are these needed now?

File: oak_client/src/transport.rs [hunk 3]
```diff
@@ -62,17 +72,20 @@ impl micro_rpc::AsyncTransport for GrpcStreamingTransport {
             ));
         };
 
-        Ok(invoke_response.encrypted_body)
+        // Ok(invoke_response.encrypted_response)
```
Comment (tiziano88) [inline oak_client/src/transport.rs:75]: Remove?

File: oak_functions_launcher/build.rs [hunk 4]
```diff
@@ -36,10 +37,7 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
             "{}oak_functions_service/proto/oak_functions.proto",
             env!("WORKSPACE_ROOT")
         )],
-        &[format!(
-            "{}oak_functions_service/proto",
-            env!("WORKSPACE_ROOT")
-        )],
+        &[env!("WORKSPACE_ROOT")],
```
Comment (tiziano88) [inline oak_functions_launcher/build.rs:40]: Did you confirm that this is not rebuilt for every unrelated change in the workspace?

File: oak_remote_attestation/proto/v1/messages.proto [hunk 5]
```diff
@@ -97,19 +99,15 @@ message GetPublicKeyResponse {
...
 message InvokeRequest {
-  // Body of the request, encrypted with the enclave public key.
-  // TODO(#3442): Specify encryption format.
-  // TODO(#3442): Also pass client public or symmetric key to use for response encryption. Note that
-  // needs to be at least authenticated by the client, if not encrypted within this field.
-  // TODO(#3843): Use explicit `oak_crypto` messages.
-  bytes encrypted_body = 1;
+  // Body of the request, encrypted using Hybrid Public Key Encryption (HPKE).
+  // <https://www.rfc-editor.org/rfc/rfc9180.html>
+  oak.crypto.v1.EncryptedRequest encrypted_request = 1;
```
Comment (tiziano88) [inline oak_remote_attestation/proto/v1/messages.proto:104]: So is this a breaking change in terms of proto compatibility? i.e. do we need to update binaries and protos and libraries at the same time?

File: java/src/main/java/com/google/oak/crypto/DecryptionResult.java (hunk 1 above)
Comment (ipetr0v) [reply java/src/main/java/com/google/oak/crypto/DecryptionResult.java:19]: `DecryptionResult` is pre-existing, it just was moved to a separate file.
Will use it as an input argument in a separate PR.

File: oak_remote_attestation/proto/v1/messages.proto (hunk 5 above)
Comment (tiziano88) [reply oak_remote_attestation/proto/v1/messages.proto:104]: Ok, there is probably some way of doing it incrementally, but up to you. In the future it may be useful to do these changes without breaking compatibility though.

Comment (tiziano88) [reply oak_remote_attestation/proto/v1/messages.proto:104]: It must always be possible to do non-breaking changes, according to the normal proto guidelines. in this case, you could add another field (e.g. with number 2) for the request, then update the libraries to use either field (or both fields together), then import those changes, then deprecate the original field. Makes sense? We cannot keep making breaking changes forever :)

Comment (tiziano88) [reply oak_remote_attestation/proto/v1/messages.proto:104]: (no need to do it here, but I want to make sure we are on the same page, i.e. that currently we are doing breaking changes, but it's possible to do them in a non-breaking way instead, although more complex)

Comment (pmcgrath17) [reply oak_remote_attestation/proto/v1/messages.proto:104]: Perhaps keep track of the Tag number at the top of the proto, continue to add proto fields, and mark old ones as deprecated when they are no longer of use.

File: Cargo.lock
Comment (pmcgrath17) [inline Cargo.lock:1]: So this is just a question out of curiosity more than anything, but whenever local changes happen to this file should we always push the change? Perhaps I am not understanding what the purpose is of the ```Cargo.lock``` file. Here and in the other ```Cargo.lock``` file.

File: cc/crypto/client_encryptor.h [hunk 6]
```diff
@@ -54,18 +55,15 @@ class ClientEncryptor {
   // Encrypts `plaintext` and authenticates `associated_data` using AEAD.
   // <https://datatracker.ietf.org/doc/html/rfc5116>
   //
-  // Returns a serialized [`oak.crypto.EncryptedRequest`] message.
-  // TODO(#3843): Return unserialized proto messages once we have Java encryption without JNI.
-  absl::StatusOr<std::string> Encrypt(absl::string_view plaintext,
-                                      absl::string_view associated_data);
+  // Returns an [`oak.crypto.EncryptedRequest`] proto message.
+  absl::StatusOr<::oak::crypto::v1::EncryptedRequest> Encrypt(absl::string_view plaintext,
```
Comment (pmcgrath17) [inline cc/crypto/client_encryptor.h:59]: NIT: as per feedback I have gotten internally, do not fully qualify the EncryptedRequest (i.e. remove the first ```::```). Only fully qualify in this case if the code fails to build. Here and and in all other relevant header files in this pull request.

File: cc/transport/grpc_streaming_transport.cc [hunk 7]
```diff
@@ -59,11 +63,11 @@ absl::StatusOr<AttestationBundle> GrpcStreamingTransport::GetEvidence() {
...
 }
 
-absl::StatusOr<std::string> GrpcStreamingTransport::Invoke(absl::string_view request_bytes) {
+absl::StatusOr<EncryptedResponse> GrpcStreamingTransport::Invoke(
+    const EncryptedRequest& encrypted_request) {
   // Create request.
   RequestWrapper request;
-  InvokeRequest* invoke_request = request.mutable_invoke_request();
-  invoke_request->set_encrypted_body(request_bytes);
+  *request.mutable_invoke_request()->mutable_encrypted_request() = encrypted_request;
```
Comment (pmcgrath17) [inline cc/transport/grpc_streaming_transport.cc:70]: More of an open-ended question here, but how large do we expect requests to be? Is copying ```encrypted_request``` here acceptable, or should we instead have the invoke function use the direct value of ```encrypted_request```?

**Review by pmcgrath17 — APPROVED** [review]
Looking over this, I think this change really helps with readability of the code for future users/developers. Nice work.

File: cc/transport/grpc_streaming_transport.cc (hunk 7 above)
Comment (ipetr0v) [reply cc/transport/grpc_streaming_transport.cc:70]: We don't limit the size of a message, so technically it could be max proto size (which is 2Mb I believe).

Though `Invoke` is async, so we probably shouldn't rely on the argument lifetime to use it as a pointer.

File: oak_remote_attestation/proto/v1/messages.proto (hunk 5 above)
Comment (tiziano88) [reply oak_remote_attestation/proto/v1/messages.proto]: I'm confused -- is the old field used anywhere at all now?

Comment (ipetr0v) [reply oak_remote_attestation/proto/v1/messages.proto]: It's only used internally

Comment (ipetr0v) [reply oak_remote_attestation/proto/v1/messages.proto:104]: So I decided to not keep the original proto field and make it a breaking change. But before submitting I will make a test internal update to see that no tests are failing.

File: oak_functions_launcher/build.rs (hunk 4 above)
Comment (ipetr0v) [reply oak_functions_launcher/build.rs:40]: Not sure, I think it still recompiles

I.e. `oak_client` crate is being built in `0.12s`, but when I change an unrelated `oak_functions.proto` it becomes `0.21s`

Comment (tiziano88) [reply oak_functions_launcher/build.rs:40]: Please make sure it does not recompile everything, and check properly (not just based on time). You can look at the output of cargo build and check whether it rebuilds this package or not.

Comment (ipetr0v) [reply oak_functions_launcher/build.rs:40]: Looks like it doesn't recompile:
```bash
$ cargo build --package=oak_client --release
    Finished release [optimized] target(s) in 0.13s
$ cargo build --package=oak_client --release
    Finished release [optimized] target(s) in 0.15s
```
```bash
$ cargo build --package=oak_functions_service
    Finished dev [unoptimized + debuginfo] target(s) in 0.13s
$ cargo build --package=oak_functions_service
    Finished dev [unoptimized + debuginfo] target(s) in 0.13s
```
